tools/cldr-code/src/main/java/org/unicode/cldr/util/NameGetter.java

```diff
@@ -0,0 +1,547 @@
+import com.ibm.icu.text.Transform;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class NameGetter {
+
+    private final CLDRFile cldrFile;
+
+    /**
+     * Construct a new NameGetter.
+     *
+     * @param cldrFile must not be null
+     */
+    public NameGetter(CLDRFile cldrFile) {
+        if (cldrFile == null) {
+            throw new RuntimeException("NameGetter must have non-null CLDRFile");
+        }
+        this.cldrFile = cldrFile;
+    }
+
+    private static final String GETNAME_LOCALE_SEPARATOR =
+            "//ldml/localeDisplayNames/localeDisplayPattern/localeSeparator";
+    private static final String GETNAME_LOCALE_PATTERN =
+            "//ldml/localeDisplayNames/localeDisplayPattern/localePattern";
+    private static final String GETNAME_LOCALE_KEY_TYPE_PATTERN =
+            "//ldml/localeDisplayNames/localeDisplayPattern/localeKeyTypePattern";
+
+    private static final Joiner JOIN_HYPHEN = Joiner.on('-');
+    private static final Joiner JOIN_UNDERBAR = Joiner.on('_');
+
+    /**
+     * Get a name, given a type (as a String) and a code.
+     *
+     * <p>Ideally getNameFromTypestrCode and getNameFromTypenumCode should be replaced by a method
+     * that takes an enum rather than String or int. Compare StandardCodes.CodeType, and the
+     * integers defined in the vicinity of CLDRFile.TERRITORY_NAME
+     *
+     * @param type a string such as "currency", "language", "region", "script", "territory", ...
+     * @param code a code such as "JP"
+     * @return the name
+     */
+    public String getNameFromTypestrCode(String type, String code) {
+        return getNameFromTypenumCode(CLDRFile.typeNameToCode(type), code);
+    }
+
+    /**
+     * Get a name, given a type (as a int) and a code.
+     *
+     * @param type the type such as CLDRFile.TERRITORY_NAME
+     * @param code a code such as "JP"
+     * @return the name
+     */
+    public String getNameFromTypenumCode(int type, String code) {
+        return getNameFromTypeCodeTransformPaths(type, code, null, null);
+    }
+
+    public String getNameFromTypeCodePaths(int type, String code, Set<String> paths) {
+        return getNameFromTypeCodeTransformPaths(type, code, null, paths);
+    }
+
+    public String getNameFromTypeCodeAltpicker(
+            int type, String code, Transform<String, String> altPicker) {
+        return getNameFromTypeCodeTransformPaths(type, code, altPicker, null);
+    }
+
+    /**
+     * Returns the name of the given bcp47 identifier. Note that extensions must be specified using
+     * the old "\@key=type" syntax.
+     *
+     * @param localeOrTZID the bcp47 identifier (locale or timezone ID)
+     * @return the name of the given bcp47 identifier
+     */
+    public synchronized String getNameFromBCP47(String localeOrTZID) {
+        return getNameFromBCP47Bool(localeOrTZID, false);
```

Comment:
As long as we are writing new code, best to avoid having a boolean in any API, and rather use a meaningful enum. Again, could be done in later PR.

Reply:
Good idea. I'll include that in a subsequent PR